WreckEngine/UI.cpp

```diff
@@ -0,0 +1,402 @@
+        ImGuiIO& io = ImGui::GetIO();
+        io.KeyMap[ImGuiKey_Tab] = GLFW_KEY_TAB;
+        io.KeyMap[ImGuiKey_LeftArrow] = GLFW_KEY_LEFT;
+        io.KeyMap[ImGuiKey_RightArrow] = GLFW_KEY_RIGHT;
+        io.KeyMap[ImGuiKey_UpArrow] = GLFW_KEY_UP;
+        io.KeyMap[ImGuiKey_DownArrow] = GLFW_KEY_DOWN;
+        io.KeyMap[ImGuiKey_PageUp] = GLFW_KEY_PAGE_UP;
+        io.KeyMap[ImGuiKey_PageDown] = GLFW_KEY_PAGE_DOWN;
+        io.KeyMap[ImGuiKey_Home] = GLFW_KEY_HOME;
+        io.KeyMap[ImGuiKey_End] = GLFW_KEY_END;
+        io.KeyMap[ImGuiKey_Delete] = GLFW_KEY_DELETE;
+        io.KeyMap[ImGuiKey_Backspace] = GLFW_KEY_BACKSPACE;
+        io.KeyMap[ImGuiKey_Enter] = GLFW_KEY_ENTER;
+        io.KeyMap[ImGuiKey_Escape] = GLFW_KEY_ESCAPE;
+        io.KeyMap[ImGuiKey_A] = GLFW_KEY_A;
+        io.KeyMap[ImGuiKey_C] = GLFW_KEY_C;
+        io.KeyMap[ImGuiKey_V] = GLFW_KEY_V;
+        io.KeyMap[ImGuiKey_X] = GLFW_KEY_X;
+        io.KeyMap[ImGuiKey_Y] = GLFW_KEY_Y;
+        io.KeyMap[ImGuiKey_Z] = GLFW_KEY_Z;
+
+        // Alternatively you can set this to NULL and call ImGui::GetDrawData() after ImGui::Render() to get the same ImDrawData pointer.
+        io.RenderDrawListsFn = ImGuiRenderImpl;
+        io.SetClipboardTextFn = ImGuiSetClipboardImpl;
+        io.GetClipboardTextFn = ImGuiGetClipboardImpl;
+        io.ClipboardUserData = Window::window;
+#ifdef _WIN32
+        io.ImeWindowHandle = glfwGetWin32Window(Window::window);
+#endif
+
+        glfwSetMouseButtonCallback(Window::window, GlfwMouseButtonCallback);
+        glfwSetScrollCallback(Window::window, GlfwScrollCallback);
+        glfwSetKeyCallback(Window::window, GlfwKeyCallback);
+        glfwSetCharCallback(Window::window, GlfwCharCallback);
+
+        return true;
+    }
+
+    /**
+     * \brief Attempts to initialize the UI system.
+     *
+     * \return True if the UI system was successfully initialized, otherwise false.
+     */
+    bool Initialize()
+    {
+        // Let's attempt to load a better looking font
+        ImGuiIO& io = ImGui::GetIO();
+        ImFont* font = nullptr;
+        io.Fonts->ClearFonts();
+        //font = io.Fonts->AddFontDefault();
+        //font = io.Fonts->AddFontFromFileTTF("Assets/Fonts/Cousine-Regular.ttf", 15.0f);
+        font = io.Fonts->AddFontFromFileTTF("Assets/Fonts/DroidSans.ttf", 12.0f);
+        //font = io.Fonts->AddFontFromFileTTF("Assets/Fonts/ProggyClean.ttf", 13.0f);
+        //font = io.Fonts->AddFontFromFileTTF("Assets/Fonts/ProggyTiny.ttf", 10.0f);
+
+        if (!font)
+        {
+            return false;
+        }
+
+        return InitializeSystem() && InitializeGraphics();
+    }
+    
+    /**
+     * \brief Prepares to draw a new frame's worth of UI.
+     */
+    void PrepareFrame()
+    {
+        ImGuiIO& io = ImGui::GetIO();
+
+        // Setup display size (every frame to accommodate for window resizing)
+        int w, h;
+        int display_w, display_h;
+        glfwGetWindowSize(Window::window, &w, &h);
```

Comment:
currently it's implemented as an update function. tbh that was lazy and should be rewritten as a resize event. but more or less yes